### ISSUE TITLE
Inefficient idle connections re-use (nbthread > 1, server maxconn set, http-reuse always)

### ISSUE THREAD

**Issue by tqinli** [haproxy/haproxy#1281]
## Detailed description of the problem

We are using haproxy as reverse proxy in front of our real-time API service that processes >10k rps. The API service is written in C#, which performs poorly if there are too many connections (>1000 per second) made to the service. Based on our requirements, we turned on the following haproxy options:

- http-reuse always
- maxconn limit set to a small number, e.g. 500
- nbthread set to a value > 1

Sometimes after a short network outage, those 10k requests/s becomes 10k connections/s against the haproxy in front of our service. However, instead of shielding most of the incoming connections by using pooled idle connections under the maxconn limit to connect to our service, haproxy is connecting/disconnecting to the backend at 7k connections/s rate, while appears to be obeying the maxconn limit we configured.

We are using haproxy v2.1, and this issue appears to be fixed in v2.2.5 by this [commit](https://github.com/haproxy/haproxy/commit/5f1ded5629), although judging by the comment, it is unclear that we fully understand what has been fixed. Also created this issue because I can't find similar issues discussed here, and maybe it would be useful in case others are facing similar problems. 

## Expected behavior

We expect if http-reuse is set to always, and maxconn is set to a value, haproxy will only create new connections when current connection < maxconn, otherwise will wait for a connection to become ilde and use it. This also means when frontend is receiving 10k connections/s, the backend should still see very low connection rate.

## Steps to reproduce the behavior

Our repro setup:
1. the client, it is a simple python3 client that opens a socket, sends a POST request and closes the socket, and repeat 200 times. We later found out the `s.close()` isn't really important to the repro at all

```
#!/usr/bin/python3
import socket, struct, time

for i in range(200):
    print(i)
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.connect(("127.0.0.1", 80))
        s.sendall(b'POST / HTTP/1.1\r\nHost: asdf\r\nContent-Length: 1000000\r\nConnection: keep-alive\r\n\r\n' + bytes(1000000*'a', 'ascii'))
        s.close()
```

2. the backend server we use `httpterm`, we simply let it listen on :8080 and enables debug logging that prints socket accept and close events.

3. haproxy that listens on 80 and forwards requests to 8080, which had maxconn set to 20, nbthread set to 16 (or rather any value greater than 1)

After we ran above `repro.py` client, which only took ~2s, we will see the following server output from `httpterm`:

```
httpterm_1      | 0000002d:dummy.accept(0005)=001f from [192.168.96.3:37108]
httpterm_1      | 0000002d:dummy.clihdr[001f:ffff]: POST / HTTP/1.1
httpterm_1      | 0000002d:dummy.clihdr[001f:ffff]: host: asdf
httpterm_1      | 0000002d:dummy.clihdr[001f:ffff]: content-length: 1000000
...
httpterm_1      | 00000022:dummy.closed[0028:ffff]
```

What we are interested are only the number of occurrences of `accept` in the output, we expected to see at most 20, because we set `maxconn` 20, and the repro was just seconds so none of the timeout should kick in and close connection unexpectedly. Instead 20, we see 155:

```
$ cat httpterm.log | grep accept | wc -l
155
```

We also ran our repro against v2.2.5 and later and found only 20 expected `accept` (though sometimes 21, let's keep it out of scope for this issue here).

## Do you have any idea what may have caused this?

Using the same repro, we triangulated and isolated this commit that fixed the problem in v2.2.5 https://github.com/haproxy/haproxy/commit/5f1ded5629 

However the comment in the fix doesn't appear to explain this issue. My guess was when a server reaches its maxconn, and a request enters `connect_server()`, if we missed one loop iteration to find idle connection in `conn_backend_get`, there is a chance we will miss finding the idle connection if it happens to be owned by the last thread we missed. Then we will skip connection re-use logic, and later we might incorrectly kill an idle connection when trying to make a new connection. But this doesn't explain why it is happening so frequently, e.g. haproxy would make 7k backend connections for every 10k frontend connection we saw.

## Do you have an idea how to solve the issue?

We are preparing to upgrade to latest v2.2 which has the fix.

## What is your configuration?

```
global
    nbproc 1
    nbthread 16

defaults
    timeout client 120s
    timeout server 15s
    timeout connect 15s
    http-reuse always

frontend external
    mode http
    bind *:80

    default_backend internal

backend internal
    mode http
    option http-buffer-request
    server s1 localhost:8080 maxconn 20
```

## Output of `haproxy -vv` and `uname -a`

```
HA-Proxy version 2.1.11-9da7aab 2021/01/08 - https://haproxy.org/
Status: stable branch - will stop receiving fixes around Q1 2021.
Known bugs: http://www.haproxy.org/bugs/bugs-2.1.11.html
Running on: Linux 4.15.0-143-generic #147-Ubuntu SMP Wed Apr 14 16:10:11 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wno-implicit-fallthrough -Wno-stringop-overflow -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2=1 USE_PCRE2_JIT=1 USE_GETADDRINFO=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1
Feature list : +EPOLL -KQUEUE -MY_EPOLL -MY_SPLICE +NETFILTER -PCRE -PCRE_JIT +PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED -REGPARM -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H -VSYSCALL +BACKTRACE +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM -MY_ACCEPT4 +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL -SYSTEMD -OBSOLETE_LINKER +PRCTL +THREAD_DUMP -EVPORTS
Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200
Built with multi-threading support (MAX_THREADS=64, default=32).
Built with OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
Running on OpenSSL version : OpenSSL 1.1.1d  10 Sep 2019
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with network namespace support.
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.32 2018-09-10
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with the Prometheus exporter as a service
Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.
Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2
            fcgi : mode=HTTP       side=BE        mux=FCGI
       <default> : mode=HTTP       side=FE|BE     mux=H1
       <default> : mode=TCP        side=FE|BE     mux=PASS
Available services :
        prometheus-exporter
Available filters :
        [SPOE] spoe
        [CACHE] cache
        [FCGI] fcgi-app
        [TRACE] trace
        [COMP] compression
```

```
Linux host 4.15.0-143-generic #147-Ubuntu SMP Wed Apr 14 16:10:11 UTC 2021 x86_64 GNU/Linux
```

## If HAProxy crashed: Last outputs and backtraces

```
no crashes
```

## Additional information (if helpful)

I was also looking into 2.1 release and found the fix isn't applicable to 2.1, as far as I can tell, which makes we wonder if the issue in 2.1 and pre 2.2.5 are the same issue, or rather slightly different issues that have different implementations that have the same symptom.

**Comment by TimWolla** [comment]
> We are preparing to upgrade to latest v2.2 which has the fix.

You already mentioned that you are planning to upgrade, but I'd still like to point out that HAProxy 2.1 *already is* EOL'd since the middle of March. Odd-numbered versions are only supported for a short-ish period of time and are meant for power users. If you need preparations to upgrade HAProxy in your infrastructure, then I recommend to stay on the even-numbered LTS version.

**Comment by tqinli** [comment]
yeah, I was not aware that odd-number version receives a shorter support period when I chose the version. Will stick to a even-number LTS version in the future :-)

**Comment by wtarreau** [comment]
In addition to what Tim said, it's important to know that idle conns were seriously revamped between 2.1 and 2.2 and for example the rules deciding when to keep then changed to be more efficient. There were further improvements in 2.4 where all attributes of the connection are used to decide whether they can be reused or not. So since you're upgrading, you might even want to consider giving a try to 2.4 :-)

I think we can close as wontfix, at least it will remain as a detailed reference for those experiencing the same issue, and with a suggested fix.

**Comment by wtarreau** [comment]
Well in fact it's "fixed" (since it was fixed later), but I'm leaving "2.1" indicating it's not fixed there :-)
